src/ui.rs

```diff
@@ -141,47 +226,130 @@ impl<'a> State<'a> {
+                Right => ProcessAction::new(Action::Pan(PanAction::Right), state.repeat).into(),
+                Up => ProcessAction::new(Action::Pan(PanAction::Up), state.repeat).into(),
+                Down => ProcessAction::new(Action::Pan(PanAction::Down), state.repeat).into(),
+                _ => MultiNormalAction::Noop,
+            },
+            (k, Mod::NOMOD) | (k, _) => match k {
+                Delete => ProcessAction::new(Action::Delete, state.repeat).into(),
+                Escape => MultiNormalAction::Cancel,
+                PageUp => ProcessAction::new(Action::SkipForward, state.repeat).into(),
+                PageDown => ProcessAction::new(Action::SkipBack, state.repeat).into(),
+                Period => ProcessAction::new(state.last_action.clone(), state.repeat).into(),
+                Right => ProcessAction::new(Action::Next, state.repeat).into(),
+                Left => ProcessAction::new(Action::Prev, state.repeat).into(),
+                Up => ProcessAction::new(Action::Zoom(ZoomAction::In), state.repeat).into(),
+                Down => ProcessAction::new(Action::Zoom(ZoomAction::Out), state.repeat).into(),
+                Backspace => ProcessAction::new(Action::Backspace, 1).into(),
+                _ => MultiNormalAction::Noop,
+            },
+        },
+
+        Event::Window { win_event, .. } => match win_event {
+            // Exposed: Rerender if the window was not changed by us.
+            Exposed | Resized(..) | SizeChanged(..) | Maximized => MultiNormalAction::ReRender,
+            _ => MultiNormalAction::Noop,
+        },
+
+        _ => MultiNormalAction::Noop,
+    }
+}
+
 /// event_action returns which action should be performed in response to this event
-pub fn process_normal_mode<'a>(state: &mut State<'a>, event: &Event) -> Action<'a> {
+pub fn process_normal_mode<'a>(state: &mut State<'a>, event: Event) -> Action<'a> {
     // Bring variants in function namespace for reduced typing.
     use sdl2::event::WindowEvent::*;
     use sdl2::keyboard::Keycode::*;
-    use sdl2::keyboard::Mod;
 
     match event {
         Event::Quit { .. } => Action::Quit,
 
         Event::TextInput { text, .. } => match text.as_str() {
-            "c" => state.process_action(Action::Copy),
-            "d" => state.process_action(Action::Delete),
-            "f" => state.process_action(Action::ToggleFullscreen),
-            "g" => state.process_action(Action::First),
-            "G" => state.process_action(Action::Last),
+            // Number of times to repeat operation
+            // 0 is not captured for first digit as it does not impact counts
+            "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" => {
+                // Safe to unwrap as only digits were matched
+                let first_digit = text.parse::<usize>().unwrap();
+                // Save the first digit before switching
+                state.repeat = first_digit;
+                Action::SwitchMultiNormalMode
```

Comment:
As stated above, maybe have the `Mode` store the number of times to repeat? (As a String or Cow maybe?) I don't know enough about Cow yet...